### DISCUSSION TITLE
Mox (QOSGrpCpuLimit)

### DISCUSSION THREAD

**Discussion by skreling** [sr320/course-fish546-2021#115] [Q&A] [answered]
I'm not sure if this is just a problem with my script or what? Trying to run a blastn command.

This is the script I was attempting to use:

```
#!/bin/bash
## Job Name
#SBATCH --job-name=546_blast_bat
## Allocation Definition
#SBATCH --account=srlab
#SBATCH --partition=srlab
## Resources
## Nodes
#SBATCH --nodes=1
## Walltime (days-hours:minutes:seconds format)
#SBATCH --time=0-12:00:00
## Memory per node
#SBATCH --mem=120G
##turn on e-mail notification
#SBATCH --mail-type=ALL
#SBATCH --mail-user=skreling@uw.edu
## Specify the working directory for this job
#SBATCH --chdir=/gscratch/scrubbed/skreling/030221_546_blast_bat


##########################################################################
# Exit script if any command fails
set -e 
#########################################################################
# Blast path
bldir = "/gscratch/srlab/programs/ncbi-blast-2.8.1+/bin/"

# Build blast database
{bldir}makeblastdb \
-in /gscratch/srlab/skreling/rererence_under.fas \
-dbtype nucl \
-out /gscratch/srlab/skreling/reference

# Run blastn
/gscratch/srlab/programs/ncbi-blast-2.8.1+/bin/blastn \
-query /gscratch/srlab/skreling/dereplicated.fasta \
-db /gscratch/srlab/skreling/reference  \
-out /gscratch/scrubbed/skreling/030221_546_blast_bat/bat_S2_blast_mox.tab \
-evalue 1E-05 \
-num_threads 40 \
-max_target_seqs 1 \
-max_hsps 1 \
-outfmt 6
```

<img width="612" alt="Screen Shot 2021-03-03 at 7 38 41 AM" src="https://user-images.githubusercontent.com/76884520/109833247-f60a3480-7bf5-11eb-8715-50781f057eff.png">

**Comment by kubu4** [comment, marked as answer]
You're all good. There are just some jobs ahead of yours. Your job will run when the others complete.

**Comment by skreling** [comment]
As an aside. It just failed, so clearly I did something wrong. Do you see anything obvious with my script?

**Comment by sr320** [comment]
No problem with the script - we only have two nodes so you are just seeing
jobs queued up because nodes (CPUlimit) has been hit. I just cancelled one
of my jobs so one of yours is now running.

**Reply by skreling** [reply]
I think there might be something wrong, because I did get an email back saying they failed. I knew the first 2 were likely to fail, but the third one (the script above) I thought would work. It's also no longer in the queue so leads me to believe something is incorrect.

**Reply by skreling** [reply]
never mind I can't spell. I spelled one of the file names wrong :(

**Comment by skreling** [comment]
Ok I don't know what the issue is now. It just keeps failing and I've double checked all of the file paths/names and I think they're all correct now. Definitely had some errors in the before.

```
#!/bin/bash
## Job Name
#SBATCH --job-name=546_blast_bat
## Allocation Definition
#SBATCH --account=srlab
#SBATCH --partition=srlab
## Resources
## Nodes
#SBATCH --nodes=1
## Walltime (days-hours:minutes:seconds format)
#SBATCH --time=0-12:00:00
## Memory per node
#SBATCH --mem=120G
##turn on e-mail notification
#SBATCH --mail-type=ALL
#SBATCH --mail-user=skreling@uw.edu
## Specify the working directory for this job
#SBATCH --chdir=/gscratch/scrubbed/skreling/030221_546_blast_bat


##########################################################################
# Exit script if any command fails
set -e
#########################################################################
# Blast path
bldir = "/gscratch/srlab/programs/ncbi-blast-2.8.1+/bin/"

# Build blast database
{bldir}makeblastdb \
-in /gscratch/scrubbed/skreling/reference_under.fas \
-dbtype nucl \
-out /gscratch/scrubbed/skreling/reference

# Run blastn
{bldir}blastn \
-db /gscratch/scrubbed/skreling/reference \
-query /gscratch/scrubbed/skreling/dereplicated.fasta \
-out /gscratch/scrubbed/skreling/030221_546_blast_bat/bat_S2_blast_mox.tab \
-outfmt 6
```

**Reply by kubu4** [reply]
Please paste contents of slurm output file. It's called: slurm.jobid.out

**Reply by skreling** [reply]
/var/spool/slurm/d/job1608152/slurm_script: line 26: bldir: command not found

It doesn't seem to like `bldir = "/gscratch/srlab/programs/ncbi-blast-2.8.1+/bin/"` but why? Is it the spaces?

**Reply by kubu4** [reply]
You can't have spaces. Needs to be:

`bldir="/gscratch/srlab/programs/ncbi-blast-2.8.1+/bin/"`

A string followed by a space tells bash that it should run that command. So, with the spaces, it was expecting a command called `bldir`.

**Reply by skreling** [reply]
Oh okay. That makes sense. I was just looking at the equivalent I had in Jupyter but that would've been python so that makes more sense.

Fixed that, but it still isn't seeming to like it, because its now giving me `/var/spool/slurm/d/job1608168/slurm_script: line 30: {bldir}makeblastdb: command not found"` I double checked the file path to the program and makeblastdb is definitely within it

**Reply by skreling** [reply]
Not sure what was wrong, but I just converted the relative file paths to absolute paths and *it appears* to have not failed yet

**Reply by kubu4** [reply]
A bit surprised it's working.

This should be a problem:

`{bldir}`

It should be prefaced by a dollar sign for bash to know that it's a variable:

`${bldir}`

**Reply by skreling** [reply]
That would make sense. It's working now because I replaced the {bldir} with the absolute file path itself
